### PR TITLE
damldocs: Document template instances.

### PR DESCRIPTION
**Pull request by associahedron**
I was most of the way to adding template instance docs to damldoc, when I realized my approach to detecting template instances did not work, and in fact *there is no way to detect that a newtype comes from a template instance* under the current generic templates desugaring proposal.

I created the issue #2239 to add a marker in the generated newtype's field docs that will let us detect generation during template instance desugaring. So the template instance support in this PR is somewhat speculative, but it should be easy to change again should template instances be desugared in a different way from the linked issue.

Nevertheless I added a test that uses the proposed marker.

This PR rounds out all the damldocs support needed for generic templates (#1387).

### CODE REVIEW

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs [hunk 1]
```diff
@@ -67,16 +67,23 @@ mkDocs opts fp = do
...
 
             md_name = dc_modname
             md_anchor = Just (moduleAnchor md_name)
             md_descr = modDoc dc_tcmod
-            md_templates = getTemplateDocs ctx typeMap templateInstanceMap
+            md_templates = getTemplateDocs ctx typeMap templateInstanceClassMap
             md_functions = mapMaybe (getFctDocs ctx) dc_decls
-            md_adts
+
+            filteredAdts -- all ADT docs without templates or choices
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: `without`? This looks more like it only has `dc_templates` and `dc_choices`.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs [hunk 2]
```diff
@@ -67,16 +67,23 @@ mkDocs opts fp = do
...
                 = MS.elems . MS.withoutKeys typeMap . Set.unions
-                $ dc_templates : MS.keysSet dc_templateInstances : MS.elems dc_choices
+                $ dc_templates : MS.elems dc_choices
+
+            (md_adts, md_templateInstances) =
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: `md_adts` does not contain all ADTs right? It looks like it only contains template and choice ADTs that are not template instances?

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs [hunk 3]
```diff
@@ -388,29 +391,51 @@ getTemplateDocs DocCtx{..} typeMap templateInstanceMap =
...
+-- @newtype ProposalIou = MkProposalIou with unProposalIou : Proposal Iou -- ^ TEMPLATE_INSTANCE@
+--
+-- So the goal of this function is to extract the template instance doc
+-- from the newtype doc if it exists.
+getTemplateInstanceDoc :: ADTDoc -> Maybe TemplateInstanceDoc
+getTemplateInstanceDoc adt
+    | ADTDoc{..} <- adt
+    , [RecordC{..}] <- ad_constrs
+    , [FieldDoc{..}] <- ac_fields
+    , Just (DocText "TEMPLATE_INSTANCE") <- fd_descr
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: Might be a good idea to add a comment here that this marker does not exist so far.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Markdown.hs [hunk 4]
```diff
@@ -89,6 +90,14 @@ tmpl2md TemplateDoc{..} = withAnchorTag td_anchor $ \tag -> mconcat
             ]
         ]
 
+renderTemplateInstanceDocAsMarkdown :: TemplateInstanceDoc -> RenderOut
+renderTemplateInstanceDocAsMarkdown TemplateInstanceDoc{..} = mconcat
+    [ renderAnchorInfix "**template instance " ti_anchor $
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Markdown.hs]: Might be a good idea to add a `bold x = "**" <> x <> "**"` helper.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs [hunk 5]
```diff
@@ -40,50 +41,33 @@ renderSimpleRst ModuleDoc{..}
...
-         ]
-  ]
+renderSimpleRst ModuleDoc{..} = mconcat
+    [ renderAnchor md_anchor
+    , renderLines
+        [ title
+        , T.replicate (T.length title) "-"
+        , maybe "" docTextToRst md_descr
+        ]
+    , section "Templates" tmpl2rst md_templates
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs]: Nice cleanup!

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs [hunk 6]
```diff
@@ -99,6 +83,17 @@ tmpl2rst TemplateDoc{..} = mconcat $
   , renderLine ""
   ] ++ map (renderIndent 2 . choiceBullet) td_choices
 
+renderTemplateInstanceDocAsRst :: TemplateInstanceDoc -> RenderOut
+renderTemplateInstanceDocAsRst TemplateInstanceDoc{..} = mconcat
+    [ renderAnchor ti_anchor
+    , renderLinesDep $ \env ->
+        [ "template instance " <> enclosedIn "**" (unTypename ti_name)
```
Comment (cocreature) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs]: Or even better `bold = enclosedIn "**"` :)

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs (hunk 1 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: It is without :-) that's what `withoutKeys` does

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs (hunk 3 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: Good idea :-) thanks

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs]: Apparently I can’t read :slightly_smiling_face:

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs (hunk 6 above)
Comment (associahedron) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Rst.hs]: a bold suggestion :-)

done

**Pull request merged**